doc/man/man4/lpp.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/lpp.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH LPP 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B lpp
\- Lightweight Presentation Protocol (LPP)
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <stropts.h>
.HP
.IB lpp_stream " = open(" /dev/tp4 ", O_RDWR);
.HP
.BI ioctl( lpp_stream ", I_PUSH, " \(lqlpp\(rq );
.\"
.\"
.SH DESCRIPTION
.PP
.B LPP
is an implementation of the
.IR "Lightweight Presentation Protocol (LPP)" ,
as described in
RFC 1085.
.[
rfc1085
.]
.B LPP
is used primarily by CMOT,
.BR cmot (4),
as described in
RFC 1095
.[
rfc1095
.]
and then
RFC 1189.
.[
rfc1189
.]
.B LPP
provides a mechanism for transporting presentation layer PDUs over TCP or UDP
(with serious cautions about using UDP for such things).
.PP
.B LPP
can be implemented as a
.I STREAMS
pushable module that is pushed over a TCP (tpi) stream.  The resulting Stream
can still present a TPI interface, just a TPI-mOSI interface as described in
XNS 5.2.
.[
xns98
.]
.\"
.\"
.SH USAGE
.PP
.B LPP
can be pushed over an OSI Transport Stream conforming to the Transport
Provider Interface (TPI),
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7),
to transform it into a Lightweight Presentation Protocol Stream.  Once pushed,
the Stream presents the minimal-OSI interface (mOSI) described in XNS 5.2.
.[
xns98
.]
.\"
.\"
.SH INTERFACE
.PP
The interface presented by the
.B LPP
is the Transport Provider Interface (TPI),
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7).
This interface conforms to the XTI mOSI (minimal-OSI) interface using the
X/Open Transport Interface (XTI),
.[
xti revision 1.0
.]
.BR xti (3),
as described in the XNS 5.2.
.[
xns98
.]
.\"
.\"
.SH OBJECT IDENTIFIERS
.PP
The following OBJECT IDENTIFIER value is defined in RFC 1095:
.[
rfc1095
.]
.ad l
.HP
CMOT
.[
rfc1095
.]
backward compatible Application Context Name:
.br
1.3.6.1.2.1.9.1.1
.I {iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) cmot1095(1)}
.B \(lq\ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex1\(rq
.ad b
.PP
The following OBJECT IDENTIFIER values are defined in RFC 1189:
.[
rfc1189
.]
.ad l
.HP
CMIP
.[
rfc1189
.]
manager event Application Context Name:
.br
1.3.6.1.2.1.9.1.2
.I {iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) manager-event-association(2)}
.B \(lq\ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex2\(rq
.HP
CMIP
.[
rfc1189
.]
manager event monitor Application Context Name:
.br
1.3.6.1.2.1.9.1.3
.I {iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) manager-event-monitor-association(3)}
.B \(lq\ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex3\(rq
.HP
CMIP
.[
rfc1189
.]
manager monitor control Application Context Name:
.br
1.3.6.1.2.1.9.1.4
.I {iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) manager-monitor-control-association(4)}
.B \(lq\ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex4\(rq
.HP
CMIP
.[
rfc1189
.]
manager full Application Context Name:
.br
1.3.6.1.2.1.9.1.5
.I {iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) manager-full-association(5)}
.B \(lq\ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex5\(rq
.HP
CMIP
.[
rfc1189
.]
agent event Application Context Name:
.br
1.3.6.1.2.1.9.1.6
.I {iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) agent-event-association(6)}
.B \(lq\ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex6\(rq
.ad b
.\"
.\"
.SH ADDRESSES
.PP
Address encoding is according to RFC 1278.
.[
rfc1278
.]
.\"
.\"
.SH OPTIONS
.PP
Options are as provided in the XTI mOSI interface as described in XNS 5.2.
.[
xns98
.]
.\"
.\"
.\".SH PRIMITIVES
.\".PP
.\"
.\"
.SH IOCTLS
.PP
.B LPP
does not provide any input-output controls of its own.
.\"
.\"
.SH NOTICES
.PP
This module is incomplete.
.\"
.\"
.SH IMPLEMENTATION
.PP
This module is currently incomplete.
.\"
.\"
.\".SH FILES
.\".PP
.\"
.\"
.SH MODULES
.PP
.B LPP
is a pushable
.IR STREAMS (9)
module named
.RI \(lq lpp \(rq.
.\"
.\"
.\".SH DEVICES
.\".PP
.\"
.\"
.SH "SEE ALSO"
.PP
.BR ns-tcp (4),
.BR xol (4),
.BR xot (4),
.BR isot (4),
.BR itot (4),
.BR iso-ip (4),
.BR tp4 (4),
.IR ISO (4).
.\"
.\"
.SH BUGS
.PP
As the
.B LPP
module is currently incomplete, bugs are currently being designed.
.\"
.\"
.SH COMPATIBILITY
.PP
The
.B LPP
module is compatible with the XTI mOSI specification of XNS 5.2.
.[
xns98
.]
.\"
.\"
.SH CONFORMANCE
.PP
RFC 1189.
.[
rfc1189
.]
.\"
.\"
.SH HISTORY
.PP
The
.B LPP
module is new for
.BR "@PACKAGE_TITLE@" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/lpp.c
/*****************************************************************************

 @(#) File: src/modules/lpp.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/lpp.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  Lightweight Presentation Protocol (LPP) module
 */

#define _SVR4_SOURCE
#define _MPS_SOURCE
#define _SUN_SOURCE

#include <sys/os7/compat.h>

#define LPP_DESCRIP	"LPP STREAMS Module"
#define LPP_EXTRA	"Part of the OpenSS7 OSI Stack for Linux Fast-STREAMS"
#define LPP_REVISION	"OpenSS7 src/modules/lpp.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define LPP_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define LPP_DEVICE	"Provides OpenSS7 Lightweight Presentation Protocol (LPP)"
#define LPP_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define LPP_LICENSE	"GPL"
#define LPP_BANNER	LPP_DESCRIP	"\n" \
			LPP_EXTRA	"\n" \
			LPP_REVISION	"\n" \
			LPP_COPYRIGHT	"\n" \
			LPP_DEVICE	"\n" \
			LPP_CONTACT	"\n"
#define LPP_SPLASH	LPP_DEVICE	" - " \
			LPP_REVISION	"\n"

#ifdef LINUX
MODULE_AUTHOR(LPP_CONTACT);
MODULE_DESCRIPTION(LPP_DESCRIP);
MODULE_SUPPORTED_DEVICE(LPP_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(LPP_LICENSE);
#endif				/* MODULE_LICENSE */
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-lpp");
#endif				/* MODULE_ALIAS */
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* LINUX */

#define LPP_MOD_ID	CONFIG_STREAMS_LPP_MODID
#define LPP_MOD_NAME	CONFIG_STREAMS_LPP_NAME

#ifndef LPP_MOD_NAME
#define LPP_MOD_NAME	"lpp"
#endif				/* LPP_MOD_NAME */

#ifndef LPP_MOD_ID
#define LPP_MOD_ID	0
#endif				/* LPP_MOD_ID */

/*
 *  STREAMS DEFINITIONS
 *  -------------------------------------------------------------------------
 */

#define MOD_ID		LPP_MOD_ID
#define MOD_NAME	LPP_MOD_NAME
#ifdef MODULE
#define MOD_BANNER	LPP_BANNER
#else				/* MODULE */
#define MOD_BANNER	LPP_SPLASH
#endif				/* MODULE */

/*
 *  LINUX INITIALIZATION
 *  -------------------------------------------------------------------------
 */

#ifdef LINUX
unsigned short modid = MOD_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else				/* module_param */
module_param(modid, short, 0444);
#endif				/* module_param */
MODULE_PARM_DESC(modid, "Module ID for LPP module.  (0 for allocation.)");

/** lppinit - initialize LPP
  */
static __init int
lppinit(void)
{
	cmn_err(CE_NOTE, MOD_BANNER);
	return (0);
}

/** lppexit - terminate LPP
  */
static __exit void
lppexit(void)
{
	return;
}

module_init(lppinit);
module_exit(lppexit);

#endif				/* LINUX */
